refactor(compartment-mapper): Move types from jsdoc to d.ts

kriskowal opened the pull request:
This refactor converts the JSDoc types in Compartment Mapper with TypeScript definitions. It also takes liberty to use our new trick for omitting types from bundles, but leaving continuity through importing types from JavaScript sources.

Much of this conversion was automatic, except `internal.d.ts` and `external.d.ts` are manually converted. Much of the motivation for this change was to take advantage of the expressivity of `.d.ts` over jsdoc typescript documents, allowing option groups to extend each other without otherwise useless intermediate names. The motivation for that consolidation, in turn, was to ensure that certain symmetries are preserved. For example, execute options are threaded through all functions that ultimately execute. I was looking specifically at the `languageForExtensions` option, which we misplaced on the `import` path when it belonged down the `mapNodeModulesPath`. That fix comes separately.

kriskowal commented on packages/compartment-mapper/src/archive-lite.js:
```diff
@@ -306,7 +310,7 @@ export const makeArchiveCompartmentMap = (compartmentMap, sources) => {
  * @param {ReadFn | ReadPowers} powers
  * @param {CompartmentMapDescriptor} compartmentMap
- * @param {ArchiveOptions} [options]
+ * @param {ArchiveLiteOptions} [options]
```
This is a bugfix. The type for ArchiveLiteOptions is narrower.

kriskowal commented on packages/compartment-mapper/src/import-hook.js:
```diff
@@ -726,11 +727,9 @@ export function makeImportNowHookMaker(
-        );
+        // This hook is responsible for ensuring that the moduleSpecifier
+        // actually refers to an exit module.
+        const exitRecord = exitModuleImportNowHook(moduleSpecifier);
```
Bugfix revealed by typing.

kriskowal replied:
@boneskull I’ve posted a new commit to this PR that fixes the typing bug but not as reflected in this outdated comment. It turns out that `exitModuleImportHook` does require the second argument but did not provide it in all cases. So, the new commit includes the second argument in the hook type and threads it through the various call sites. Please take a look.

kriskowal commented on packages/compartment-mapper/src/map-parser.js:
```diff
@@ -19,14 +16,17 @@ import { parseExtension } from './extension.js';
  *   ModuleTransforms,
  *   ParseFn,
- *   ParseFnAsync,
+ *   AsyncParseFn,
```
I’m choosing to treat this as non-breaking since nothing in practice relies on these type names outside of the compartment mapper, although parsers are part of the public API. We tend to treat them as opaque handles and very much doubt there are parsers implemented outside Compartment Mapper.

boneskull replied:
I can live with it

kriskowal commented on packages/compartment-mapper/src/map-parser.js:
```diff
@@ -19,14 +16,17 @@ import { parseExtension } from './extension.js';
  * } from './types.js';
  */
 
+import { syncTrampoline, asyncTrampoline } from '@endo/trampoline';
+import { parseExtension } from './extension.js';
```
This moved in a fit of homogenization of the order of type import comments relative to imports.

kriskowal commented on packages/compartment-mapper/src/types/compartment-map-schema.d.ts:
```diff
@@ -0,0 +1,116 @@
+ * `package.json`, there is a corresponding module descriptor.
+ */
+export type ModuleDescriptor = {
+  compartment?: string | undefined;
+  module?: string | undefined;
+  location?: string | undefined;
+  parser?: Language | undefined;
+  /** in base 16, hex */
+  sha512?: string | undefined;
+  exit?: string | undefined;
+  deferredError?: string | undefined;
+  retained?: boolean | undefined;
```
Nit: tsc generated redundant `| undefined` here and I missed them in my first pass. Circle back.

kriskowal commented on packages/compartment-mapper/src/types/external.d.ts:
```diff
@@ -0,0 +1,352 @@
+  commonDependencies: Record<string, string>;
+  /** Maps extensions to languages for all packages, like `txt` to `text` */
+  languageForExtension: LanguageForExtension;
+  /** Maps extensions to languages for all type=module packages */
+  moduleLanguageForExtension: LanguageForExtension;
+  /** Maps extensions to languages for all type=commonjs packages (default) */
+  commonjsLanguageForExtension: LanguageForExtension;
```
Nit: this feature appears later in this stack.

kriskowal commented on packages/compartment-mapper/src/types/external.d.ts:
```diff
@@ -0,0 +1,352 @@
+// ////////////////////////////////////////////////////////////////////////////////
+// Common option groups:
+
+export type SyncOrAsyncArchiveOptions = Partial<{
+  captureSourceLocation: CaptureSourceLocationHook;
+}> &
+  ParserForLanguageOption &
+  CompartmentOption &
+  PolicyOption;
+
+type SyncOrAsyncImportingOptions = SearchSuffixesOption & SourceMapHookOption;
+
+type ImportingOptions = ModulesOption &
+  SyncOrAsyncImportingOptions &
+  ExitModuleImportHookOption;
+
+type SyncImportingOptions = ModulesOption &
+  SyncOrAsyncImportingOptions &
+  ExitModuleImportNowHookOption;
+
+type LinkingOptions = ParserForLanguageOption &
+  CompartmentOption &
+  LanguageForExtensionOption &
+  SyncModuleTransformsOption &
+  ModuleTransformsOption;
```
Much of what I want to achieve here is sensible groupings of options here, and to ensure that all of these options get threaded to their constituent subroutines. For example, `languageForExtension` should be threaded from `loadLocation`, `makeArchive`, and such all the way down to `mapNodeModules` where it’s used.

kriskowal commented on packages/compartment-mapper/src/types/external.d.ts:
```diff
@@ -0,0 +1,352 @@
+export type Sources = Record<string, CompartmentSources>;
+export type CompartmentSources = Record<string, ModuleSource>;
+
+// TODO unionize:
```
For a future change, moving from jsdoc substantially improves our chances of turning the `ModuleSource` type from a `Partial` to a union of all valid combinations. That’ll probably complicate the code, but improve our understanding of what properties are expected on module sources of various shapes. They may need a vestigial discriminating property.

kriskowal commented on packages/compartment-mapper/src/types/node-powers.d.ts:
```diff
@@ -0,0 +1,51 @@
+type Hash = {
+  update: (data: Uint8Array | string) => Hash;
+  digest: () => {
+    toString: (radix: 'hex') => string;
```
This was `Buffer`, but we just need this one behavior and `Buffer` was hard to capture from Node.js types otherwise.

turadg replied:
good choice. Buffer type has been changing in TS versions lately.

worth a comment tho

turadg commented on packages/compartment-mapper/src/types.d.ts:
```diff
@@ -0,0 +1,11 @@
+// ESlint does not understand export *
+/* eslint-disable import/export */
+
+export * from './types/external.js';
```
Since you have this indirection, all the modules below can be `.ts`. Moreover they should be, because they are pure types rather than annotations about other `.js` files.

It would also prevent ignoring them when `skipLibCheck: true`

kriskowal replied:
For the record, we reversed this decision out of band. We are enforcing the limit with the build configuration and I’ve renamed these `.ts` so that `tsc` will complain about errors in them, per yours and @boneskull’s observation. The barrel `types.d.ts` (external) and `src/types.d.ts` (internal)  remain in `.d.ts`.

boneskull commented:
Types may be omitted from generated `.d.ts` files by using the `@internal` tag and adding `stripInternal: true` to `tsconfig.json`.

boneskull commented on packages/compartment-mapper/src/import-archive-lite.js:
```diff
@@ -405,7 +401,7 @@ export const parseArchive = async (
 
     // eslint-disable-next-line dot-notation
-    return compartment['import'](moduleSpecifier);
+    return compartment['import'](entryModuleSpecifier);
```
I assume this is for evasion

kriskowal replied:
It was, though that is not necessary any more. It predates some added sophistication to the evasive transforms.

boneskull commented on packages/compartment-mapper/src/link.js:
```diff
@@ -19,13 +20,12 @@
  *   ParserImplementation,
  *   ShouldDeferError,
  * } from './types.js'
+ * @import {ERef} from '@endo/eventual-send'
```
If the type consuming `ERef` is present in an _external_ type, then `@endop/eventual-send` it should be a production dependency of `@endo/compartment-mapper`.  If not, it _should_ be added as a dev dependency anyway, but might be OK to just leave it as-is given we're in a monorepo.

boneskull commented on packages/compartment-mapper/src/node-powers.js:
```diff
@@ -138,10 +144,10 @@ const makeReadPowersSloppy = ({
       } else {
         const realPath = await fs.promises.realpath(fileURLToPath(location));
-        return pathToFileURL(realPath).toString();
+        return pathToFileURL(realPath).href;
```
So I am curious about this.  I know that we (who?) settled on using `.href` to convert a `URL to a `string`.

`URL.prototype.toString()` just returns the `href` prop.  The advantage I see to using `toString()` (or wrapping in a template literal) is that `href` is writable and might be a source of subtle bugs (e.g., accidental assignment).

kriskowal replied:
No decision has been made and this change is not necessary. I started with `toString()` but have been leaning in the direction of `href`. There isn’t a strong case either way.

boneskull commented on packages/compartment-mapper/src/search.js:
```diff
@@ -4,9 +4,13 @@
-/** @import {ReadFn} from './types.js' */
-/** @import {ReadPowers} from './types.js' */
-/** @import {MaybeReadPowers} from './types.js' */
+/**
+ * @import {
+ *   ReadFn,
+ *   ReadPowers,
+ *   MaybeReadPowers,
+ * } from './types.js'
+ */
```
I notice that in all of our `@import` blocks, we do not end each statement with a semicolon.  I am unsure if this is considered inconsistent given the presence of semicolons everywhere else

kriskowal replied:
My understanding is that `@` pragmas are not generally terminated with a semicolon so I assumed they wouldn’t even be tolerated. Would have to check. My preference is to omit them given that they are not necessary.

boneskull commented on packages/compartment-mapper/src/types.d.ts:
```diff
@@ -0,0 +1,11 @@
+export * from './types/policy.ts';
+export * from './types/powers.ts';
+export * from './types/node-powers.ts';
+export * from './types/typescript.ts';
```
These should all end in `.js` afaik?

kriskowal replied:
It works either way and I don’t have a strong preference. I originally wrote these with `.js` and @turadg recommended `.ts`. I don’t have a principled position to stand on either way.

kriskowal replied:
I’ve changed all of these to `.js`.

mhofman replied:
I still see the specifiers here using `.ts`

kriskowal replied:
This conversation is tied to an outdated delta.

boneskull commented on packages/compartment-mapper/src/types/compartment-map-schema.d.ts:
- We should probably normalize comment grammar
- None of these `.d.ts` files are type-checked, whereas `types.js` _was_ type-checked. 
 I don't like that. This can be resolved in one of two ways:
  - Set `skipLibCheck` to `false` (please don't)
  - Rename each `.d.ts` file to `.ts`. Associated `.d.ts` files will be generated for each at the same time as the `.js` files
- We should settle on `@file` vs `@module` vs `@packageDocumentation`; I think TypeDoc prefers one of the latter two

kriskowal replied:
These correspond more closely to `@module`. I expect `@packageDocumentation` to only exist once in a package. I had `@overview` and eslint suggested `@file`.

I’ll take a look.

kriskowal replied:
Added a commit that converges all the existing `@file` and `@fileoverview` pragmas on `@module`.

boneskull commented on packages/compartment-mapper/src/types/external.d.ts:
```diff
@@ -0,0 +1,350 @@
+  SyncImportingOptions &
+  ExitModuleImportNowHookOption;
+
+export type ArchiveOptions = Omit<MapNodeModulesOptions, 'language'> &
+  ArchiveLiteOptions;
+
+export type SyncArchiveOptions = Omit<MapNodeModulesOptions, 'languages'> &
+  SyncArchiveLiteOptions;
```
one of these is wrong

boneskull commented on packages/compartment-mapper/src/types/external.d.ts:
```diff
@@ -0,0 +1,350 @@
+   * bundler, and does not attempt to vary the behavior of resolution depending
+   * on the language of the importing module.
+   */
+  searchSuffixes?: string[];
```
`string[]` or `Array<string>`???

kriskowal replied:
I will change this to `Array` because that’s the prevailing convention, but really want to ignore this discrepancy unless there’s a meaningful difference and a lint rule to enforce it.

boneskull commented on packages/compartment-mapper/src/types/external.d.ts:
```diff
@@ -0,0 +1,350 @@
+  sha512: string;
+};
+
+type ModuleTransforms = Record<string, ModuleTransform>;
+
+type SyncModuleTransforms = Record<string, SyncModuleTransform>;
+
+type ModuleTransform = (
+  ...args: ModuleTransformArguments
+) => Promise<ModuleTransformResult>;
+type SyncModuleTransform = (
+  ...args: ModuleTransformArguments
+) => ModuleTransformResult;
+
+type ModuleTransformArguments = [
+  bytes: Uint8Array,
+  specifier: string,
+  moduleLocation: string,
+  packageLocation: string,
+  params: {
+    sourceMap?: string;
+  },
+];
+
+type ModuleTransformResult = {
+  bytes: Uint8Array;
+  parser: Language;
+  sourceMap?: string;
+};
```
I think all of these should be exported, since the `moduleTransforms` etc. options are exposed

boneskull commented on packages/compartment-mapper/src/types/external.d.ts:
```diff
@@ -0,0 +1,350 @@
+  sourceLocation: string,
+) => void;
+
+export type ComputeSourceMapLocationHook = (details: {
+  compartment: string;
+  module: string;
+  location: string;
+  sha512: string;
+}) => string;
```
Extract the options object to its own type?

boneskull commented on packages/compartment-mapper/src/types/internal.d.ts:
```diff
@@ -0,0 +1,218 @@
+  parse: ParseFn | AsyncParseFn;
+  compartments: Record<string, Compartment>;
+  // TODO why are these absent? Why this asymmetry?
+  // attenuators: DeferredAttenuatorsProvider,
+  // shouldDeferError: ShouldDeferError,
```
they are ignored/unused

boneskull replied:
...and maybe we should consider adding `attenuators: never, shouldDeferError: never` to this type, just so whoever is using it cannot expect any behavior based on these two props

kriskowal replied:
I’m patching this as advised, and documenting the deliberate discrepancy, but is there a principled argument why they are not respected, to the extent that it would not make sense to add behavior for these properties as a future feature, or are they omitted for lack of a motivating use case?

boneskull replied:
I _thought_ I understood why `shouldDeferError` was omitted, but now I'm not so sure.  I _think_ it's essentially "too late" to defer an error.  @naugtur, can you explain this (again)?  

I'm assuming `DeferredAttenuatorsProvider` is not provided because a) there's only one attenuator compartment and b) that compartment is loaded via `importHook` instead of `importNowHook`.

boneskull commented on packages/compartment-mapper/src/types/node-powers.d.ts:
```diff
@@ -0,0 +1,52 @@
+
+/* eslint-disable no-use-before-define */
+
+/** For creating {@link ReadPowers} */
```
```suggestion
/** For creating `ReadPowers` */
```

The `@link` is now broken. There may be a few broken `@link`s in these files

kriskowal replied:
Did a holistic pass changing `@link` to bare words or ticked words if the referent was in a different file. Push forthcoming.

turadg reviewed (approved):
🚀

boneskull commented on packages/compartment-mapper/src/types/policy.d.ts:
```diff
@@ -0,0 +1,42 @@
+  path: Array<string>;
+};
+
+export type Attenuator<
+  GlobalParams extends [any, ...any[]] = [any, ...any[]],
+  ModuleParams extends [any, ...any[]] = [any, ...any[]],
+> = {
+  attenuateGlobals?: GlobalAttenuatorFn<GlobalParams> | undefined;
+  attenuateModule?:
+    | ModuleAttenuatorFn<ModuleParams, SomeObject, SomeObject>
+    | undefined;
+};
+
+export type GlobalAttenuatorFn<
+  Params extends [any, ...any[]] = [any, ...any[]],
+> = (
+  params: Params,
+  originalObject: Record<PropertyKey, any>,
+  globalThis: Record<PropertyKey, any>,
+) => void;
+
+export type ModuleAttenuatorFn<
+  Params extends [any, ...any[]] = [any, ...any[]],
+  T = SomeObject,
+  U = T,
+> = (params: Params, ns: T) => U;
```
I noticed that all of these `X extends [any, ...any[]] = [any, ...any[]]` should be `X extends readonly [any, ...any[]] = [unknown, ...unknown[]]`.

Also, if it makes sense that e.g., a global attenuator function accepts no parameters whatsoever, then that should be `X extends readonly any[] = unknown[]`.

kriskowal replied:
I’m going to leave this for a subsequent fix since the recommended fix did not trivially pass and is beyond my ken.

boneskull replied:
I've put this on my List.

boneskull commented on packages/compartment-mapper/src/types/powers.d.ts:
```diff
@@ -0,0 +1,120 @@
+
+import { SomeObject } from './typescript.ts';
+
+// Read
```
do audiobooks count?

boneskull commented on packages/compartment-mapper/src/types/typescript.d.ts:
```diff
@@ -0,0 +1,39 @@
+/**
+ * @file Helpers relevant to any TypeScript project.
```
there might be a copyright/license notice we need to include from [type-fest](https://npm.im/type-fest), as these types were copied from it.

kriskowal replied:
Thank you for calling this out. `type-fest` is CC0, so no attribution is required. I’m inclined to add a citation regardless, but doesn’t need a copy of the license.

boneskull reviewed (commented):
I love this direction, as it's easier to define these types in TS sources. The problem -- and it's a sticking point for me -- is that the `.d.ts` files are not type-checked.

boneskull reviewed (changes requested):
I love this direction, as it's easier to define these types in TS sources. The problem -- and it's a sticking point for me -- is that the `.d.ts` files are not type-checked.  

**UPDATE:** I guess the `@endo/eventual-send` thing is kind of a sticking point as well.

kriskowal commented:
> I love this direction, as it's easier to define these types in TS sources. The problem -- and it's a sticking point for me -- is that the `.d.ts` files are not type-checked.

Is there some other way to enforce that these hand-written type definitions must contain only type definitions?

boneskull commented:
For type-checking, I'd strongly recommend just renaming `.d.ts` files to `.ts` files.

Using `skipLibCheck: false` makes us responsible for the correctness of the types of our dependencies--not just our own. To solve problems, we might have to do things like:
- Introduce a new dependency to fix a "missing" type only
- "Missing" types may have no actual impact on our own type fidelity (as we're not guaranteed to be actually _using_ any of those types)
- "Missing" types may have no actual impact on the type fidelity of our users (and probably won't)
- Incorrect types may need to be fixed by using module augmentation
- Incorrect types may need to be addressed via `patch-package` (or whatever `yarn` does here) . Those fixes will not cascade to our consumers unless we use lifecycle scripts

A concern is that using `.d.ts` files guarantees that the files contain _only_ types. That is virtuous--but there are other ways to enforce this. I can think of:

- It's likely ESLint can be bent in this direction (someone probably has had this problem already)
- Use `export type * from './foo'` in a barrel `index.d.ts` may work; this barrel won't be typechecked, but everything else would, I think (haven't tried this)
- PR reviews 😄 

We can also use [emitDeclarationsOnly](https://www.typescriptlang.org/tsconfig/#emitDeclarationOnly) to ensure no `.js` files are ever generated. This won't _enforce_ that the `.ts` files remain execution-free, but it will prevent compilation to JS.

boneskull commented:
Also, how else are we going to export `enum`s?  Ha ha ha.

boneskull commented:
AFAIK, by default, _all_ extensions present in _any_ `.js`, `.ts`, or `.d.ts` file should be `.js` (unless we're using `.mjs`/`.cjs`). There's a `tsconfig.json` option to change this, but I haven't tried using it and don't understand the use-case.

kriskowal commented:
We have `emitDeclarationOnly` in the relevant config, so I will take that as sufficient. I can also keep the `.d.ts` for the barrel type exports files.

So, for @turadg’s and @boneskull’s consideration, I propose that I will:
1. rename `src/types/*.d.ts` to `src/types/*.ts`.
2. readdress `*.ts` to `*.js` in all `*.ts` and `*.d.ts`.

Please reactji.

kriskowal commented:
This is ready for another pass @boneskull. I believe I’ve addressed all your feedback except for the bugfix you note for policy interfaces, which I would ask you to pursue in a follow-up since it’s orthogonal, delicate, and out of my wheelhouse.

mhofman commented on packages/compartment-mapper/types.d.ts:
This seems like a duplicate of `types-index.d.ts`. How do you propose to keep them in sync under maintenance?

I'm actually wondering if this file shouldn't just go away given that we export types through the runtime entrypoint now.

kriskowal replied:
`./src/types.d.ts` exports everything under `./src/types/*.ts` whereas `./types.d.ts` does not export the types in `internal` or `typescript` since they do not participate in the public API. I _could_ eliminate the internal barrel and break up the type imports. I would like to defer that to a future refactor.

kriskowal replied:
`./src/types.d.ts` exports everything under `./src/types/*.ts` whereas `./types.d.ts` does not export the types in `internal` or `typescript` since they do not participate in the public API. I _could_ eliminate the internal barrel and break up the type imports. I would like to defer that to a future refactor.

mhofman commented on packages/compartment-mapper/src/types.d.ts:
This seems like a superset of `../types-index.d.ts`. Should we look into making that explicit somehow by having the internal file reexport the external file? (they would likely have to be moved around)

mhofman commented on packages/compartment-mapper/src/types.js:
So we're positive all internal usages of `types.js` are through type imports and never runtime imports?

kriskowal replied:
I’m pretty sure that is the case. Can I delete this stub?

kriskowal replied:
Yes.

mhofman reviewed (commented):
I only reviewed the organization of the type files.

boneskull reviewed (approved):
LGTM. I've confirmed this works as linked against `@lavamoat/node`.  ship it

mhofman commented on packages/compartment-mapper/package.json:
```diff
@@ -21,10 +21,8 @@
-  "types": "./types.d.ts",
   "exports": {
     ".": {
-      "types": "./types.d.ts",
       "default": "./index.js"
     },
```
Technically this can be simplified, no?

```suggestion
    ".": "./index.js",
```

mhofman reviewed (approved):
LGTM
